File: app/backend/controllers/iop/transfer/impl/ExternalReservationHandlerImpl.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.controllers.iop.transfer.impl;

import akka.actor.ActorSystem;
import backend.controllers.iop.transfer.api.ExternalReservationHandler;
import backend.impl.EmailComposer;
import backend.models.ExamEnrolment;
import backend.models.Reservation;
import backend.models.User;
import backend.models.iop.ExternalReservation;
import backend.util.datetime.DateTimeUtils;
import com.fasterxml.jackson.databind.JsonNode;
import com.typesafe.config.ConfigFactory;
import io.ebean.Ebean;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import javax.inject.Inject;
import org.joda.time.DateTime;
import play.Logger;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;
import scala.concurrent.duration.Duration;

public class ExternalReservationHandlerImpl implements ExternalReservationHandler {
    @Inject
    WSClient wsClient;

    @Inject
    ActorSystem system;

    @Inject
    EmailComposer emailComposer;

    private static final Logger.ALogger logger = Logger.of(ExternalReservationHandlerImpl.class);

    private static URL parseUrl(String orgRef, String facilityRef, String reservationRef) throws MalformedURLException {
        StringBuilder sb = new StringBuilder(ConfigFactory.load().getString("sitnet.integration.iop.host"));
        sb.append(String.format("/api/organisations/%s/facilities/%s/reservations", orgRef, facilityRef));
        if (reservationRef != null) {
            sb.append("/").append(reservationRef);
        }
        return new URL(sb.toString());
    }

    @Override
    public CompletionStage<Optional<Integer>> removeExternalReservation(Reservation reservation) {
        ExternalReservation external = reservation.getExternalReservation();
        URL url;
        try {
            url = parseUrl(external.getOrgRef(), external.getRoomRef(), reservation.getExternalRef());
        } catch (MalformedURLException e) {
            return CompletableFuture.completedFuture(Optional.of(Http.Status.INTERNAL_SERVER_ERROR));
        }
        WSRequest request = wsClient.url(url.toString());
        Function<WSResponse, Optional<Integer>> onSuccess = response -> {
            if (response.getStatus() != Http.Status.OK) {
                return Optional.of(Http.Status.INTERNAL_SERVER_ERROR);
            }
            return Optional.empty();
        };
        return request.delete().thenApplyAsync(onSuccess);
    }

    private CompletionStage<Result> requestRemoval(String ref, User user, String msg) throws IOException {
        final ExamEnrolment enrolment = Ebean
            .find(ExamEnrolment.class)
            .fetch("reservation")
            .fetch("reservation.machine")
            .fetch("reservation.machine.room")
            .where()
            .eq("user.id", user.getId())
            .eq("reservation.externalRef", ref)
            .findOne();
        if (enrolment == null) {
            return CompletableFuture.completedFuture(
                Results.notFound(String.format("No reservation with ref %s for current user.", ref))
            );
        }
        // Removal not permitted if reservation is in the past or ongoing
        final Reservation reservation = enrolment.getReservation();
        DateTime now = DateTimeUtils.adjustDST(DateTime.now(), reservation.getExternalReservation());
        if (reservation.toInterval().isBefore(now) || reservation.toInterval().contains(now)) {
            return CompletableFuture.completedFuture(Results.forbidden("sitnet_reservation_in_effect"));
        }
        // good to go
        ExternalReservation external = reservation.getExternalReservation();
        URL url = parseUrl(external.getOrgRef(), external.getRoomRef(), ref);
        WSRequest request = wsClient.url(url.toString());
        Function<WSResponse, Result> onSuccess = response -> {
            if (response.getStatus() != Http.Status.OK) {
                JsonNode root = response.asJson();
                return Results.internalServerError(root.get("message").asText("Connection refused"));
            }
            enrolment.setReservation(null);
            enrolment.setReservationCanceled(true);
            Ebean.save(enrolment);
            reservation.delete();

            // send email asynchronously
            boolean isStudentUser = user.equals(enrolment.getUser());
            system
                .scheduler()
                .scheduleOnce(
                    Duration.create(1, TimeUnit.SECONDS),
                    () -> {
                        emailComposer.composeReservationCancellationNotification(
                            enrolment.getUser(),
                            reservation,
                            msg,
                            isStudentUser,
                            enrolment
                        );
                        logger.info("Reservation cancellation confirmation email sent");
                    },
                    system.dispatcher()
                );

            return Results.ok();
        };
        return request.delete().thenApplyAsync(onSuccess);
    }

    // remove reservation on external side, initiated by reservation holder
    @Override
    public CompletionStage<Result> removeReservation(Reservation reservation, User user, String msg) {
        if (reservation.getExternalReservation() == null) {
            return CompletableFuture.completedFuture(Results.ok());
        }
        try {
            return requestRemoval(reservation.getExternalRef(), user, msg);
        } catch (IOException e) {
            return CompletableFuture.completedFuture(Results.internalServerError(e.getMessage()));
        }
    }
}

File: app/features/iop/transfer/impl/ExternalReservationHandlerImpl.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.iop.transfer.impl

import features.iop.transfer.api.ExternalReservationHandler
import io.ebean.DB
import database.EbeanQueryExtensions
import models.enrolment.{ExamEnrolment, Reservation}
import models.user.User
import org.joda.time.DateTime
import play.api.Logging
import play.api.http.Status._
import play.api.libs.ws.{JsonBodyWritables, WSClient}
import play.api.mvc.Result
import play.api.mvc.Results._
import services.config.ConfigReader
import services.datetime.DateTimeHandler
import services.mail.EmailComposer

import java.net.{MalformedURLException, URI}
import javax.inject.Inject
import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future}

class ExternalReservationHandlerImpl @Inject() (
    wsClient: WSClient,
    emailComposer: EmailComposer,
    dateTimeHandler: DateTimeHandler,
    configReader: ConfigReader
)(implicit ec: ExecutionContext)
    extends ExternalReservationHandler
    with EbeanQueryExtensions
    with JsonBodyWritables
    with Logging:

  private def parseUrl(orgRef: String, facilityRef: String, reservationRef: Option[String]): URI =
    val sb = new StringBuilder(configReader.getIopHost)
    sb.append(s"/api/organisations/$orgRef/facilities/$facilityRef/reservations")
    reservationRef.foreach(ref => sb.append(s"/$ref"))
    URI.create(sb.toString)

  override def removeExternalReservation(reservation: Reservation): Future[Option[Int]] =
    val external = reservation.getExternalReservation
    try
      val url = parseUrl(external.getOrgRef, external.getRoomRef, Some(reservation.getExternalRef))
      wsClient
        .url(url.toString)
        .delete()
        .map { response =>
          if response.status != OK then Some(INTERNAL_SERVER_ERROR)
          else None
        }
        .recover { case _ =>
          Some(INTERNAL_SERVER_ERROR)
        }
    catch
      case _: MalformedURLException =>
        Future.successful(Some(INTERNAL_SERVER_ERROR))

  private def requestRemoval(ref: String, user: User, msg: String): Future[Result] =
    val enrolmentOpt =
      DB.find(classOf[ExamEnrolment])
        .fetch("reservation")
        .fetch("reservation.machine")
        .fetch("reservation.machine.room")
        .where()
        .eq("user.id", user.getId)
        .eq("reservation.externalRef", ref)
        .find

    enrolmentOpt match
      case None =>
        Future.successful(NotFound(s"No reservation with ref $ref for current user."))
      case Some(enrolment) =>
        // Removal is not permitted if the reservation is in the past or ongoing
        val reservation = enrolment.getReservation
        val now         = dateTimeHandler.adjustDST(DateTime.now(), reservation.getExternalReservation)

        if reservation.toInterval.isBefore(now) || reservation.toInterval.contains(now) then
          Future.successful(Forbidden("i18n_reservation_in_effect"))
        else
          // good to go
          val external = reservation.getExternalReservation
          try
            val url = parseUrl(external.getOrgRef, external.getRoomRef, Some(ref))
            wsClient
              .url(url.toString)
              .delete()
              .map { response =>
                if response.status != OK then
                  val msg = (response.json \ "message").asOpt[String].getOrElse("Connection refused")
                  InternalServerError(msg)
                else
                  enrolment.setReservation(null)
                  enrolment.setReservationCanceled(true)
                  DB.save(enrolment)
                  reservation.delete()

                  // send email asynchronously
                  val isStudentUser = user == enrolment.getUser
                  emailComposer.scheduleEmail(1.second) {
                    emailComposer.composeReservationCancellationNotification(
                      enrolment.getUser,
                      reservation,
                      Some(msg),
                      isStudentUser,
                      enrolment
                    )
                    logger.info("Reservation cancellation confirmation email sent")
                  }

                  Ok
              }
          catch
            case e: MalformedURLException =>
              Future.successful(InternalServerError(e.getMessage))

  // remove reservation on the external side, initiated by the reservation holder
  override def removeReservation(reservation: Reservation, user: User, msg: String): Future[Result] =
    if Option(reservation.getExternalReservation).isEmpty then Future.successful(Ok)
    else requestRemoval(reservation.getExternalRef, user, msg)
